// generator-projectConfigGenerator/app/config-options-generator-mixin.js
/**
 * Created by yevhen_chmykhun on 23.10.15.
 */
var constants = require('./constants.json');
var configOptionsGeneratorMixin = function () {
    this.configFileName = 'someConfig.json';

    this.configObject = {};
    this.models = {};
    this.relationships = {};
    this.modelsCrudMethods = {};
    this.tablesColumnsFakerType = {};
    this.onlyServerFields = {};
    this.updateModelsAllowedAttrs = {};
    this.serverConfig = {};
    this.iosConfig = {};

    this.modelsCanBeUpdated = [];
    this.modelsHaveUpdateAllowedAttrs = [];

    this.modelNames = [];
    this.pureModelNames = [];
    this.relTypes = [];

    this.tableColumnsToFake = [];

    this.modelOrientedRelations = {};

    this.minContent = false;

    this.modelNames.push("none");
    for (var relationIndex = 0; relationIndex < constants.relationshipsTypes.length; relationIndex++) {
        this.relTypes.push(constants.relationshipsTypes[relationIndex]);
        this.relationships[constants.relationshipsTypes[relationIndex]] = [];
    }
    this.relTypes.push("none");
};
module.exports = configOptionsGeneratorMixin;
