// generator-projectConfigGenerator/app/prompt-chain-generator.js
/**
 * Created by yevhen_chmykhun on 13.10.15.
 */
var generalPromptGenerator = require('./prompts/general-prompt');
var modelPromptGenerator = require('./prompts/model-prompt');
var relationshipPromptGenerator = require('./prompts/relations-prompt');
var modelsCRUDMethodsPromptGenerator = require('./prompts/models-crud-methods-prompt');
var onlyServerFieldsPromptGenerator = require('./prompts/only-server-fields-prompt');
var tablesColumnsFakerTypePromptGenerator = require('./prompts/tables-faker-type-prompt');
var updateModelsAllowedAttrsPromptGenerator = require('./prompts/update-models-allowed-attrs-prompt');
var serverConfigPromptGenerator = require('./prompts/server-config-prompt');
var iosConfigPromptGenerator = require('./prompts/ios-config-prompt');
var lastPromptGenerator = require('./prompts/last-prompt');
/**
 * @param {array} chain string array of names of prompts
 * */

var promptGenerators = {
  lastPrompt: lastPromptGenerator,
  iosConfigPrompt: iosConfigPromptGenerator,
  serverConfigPrompt: serverConfigPromptGenerator,
  updateModelsAllowedAttrsPrompt: updateModelsAllowedAttrsPromptGenerator,
  tablesColumnsFakerTypePrompt: tablesColumnsFakerTypePromptGenerator,
  onlyServerFieldsPrompt: onlyServerFieldsPromptGenerator,
  modelsCRUDMethodsPrompt: modelsCRUDMethodsPromptGenerator,
  relationsPrompt: relationshipPromptGenerator,
  modelPrompt: modelPromptGenerator,
  generalPrompt: generalPromptGenerator
};
var chainGenerator = function (generator) {
  this.generator = generator;
};
chainGenerator.prototype.generateChain = function (chain) {
  var previous = lastPromptGenerator(this);
  var current;
  try {
    for (var i = chain.length - 1; i >= 0; i--) {
      if (promptGenerators[chain[i]]) {
        current = promptGenerators[chain[i]](this.generator, previous);
        previous = current;
      }
    }
  }
  catch
      (err) {
    console.log('error during prompt chain generating:' + err);
  }
  return {
    chain: chain,
    run: function () {
      current();
    }
  };
}
;

module.exports = chainGenerator;
