// generator-projectConfigGenerator/app/prompt-chain-generator.js
/**
 * Created by yevhen_chmykhun on 13.10.15.
 */
/**
 * @param {array} chain string array of names of prompts
 * */

var promptGenerators = {
  lastPrompt: require('./prompts/last-prompt'),
  iosConfigPrompt: require('./prompts/ios-config-prompt'),
  serverConfigPrompt: require('./prompts/server-config-prompt'),
  updateModelsAllowedAttrsPrompt: require('./prompts/update-models-allowed-attrs-prompt'),
  tablesColumnsFakerTypePrompt: require('./prompts/tables-faker-type-prompt'),
  onlyServerFieldsPrompt: require('./prompts/only-server-fields-prompt'),
  modelsCRUDMethodsPrompt: require('./prompts/models-crud-methods-prompt'),
  relationsPrompt: require('./prompts/relations-prompt'),
  modelPrompt: require('./prompts/model-prompt'),
  generalPrompt: require('./prompts/general-prompt')
};
var chainGenerator = function (generator) {
  this.generator = generator;
};
chainGenerator.prototype.generateChain = function (chain) {
  var previous = promptGenerators.lastPrompt(this.generator);
  var current;
  try {
    for (var i = chain.length - 1; i >= 0; i--) {
      if (promptGenerators[chain[i]]) {
        current = promptGenerators[chain[i]](this.generator, previous);
        previous = current;
      }
    }
  }
  catch
      (err) {
    console.log('error during prompt chain generating:' + err);
  }
  return {
    chain: chain,
    run: function () {
      current();
    }
  };
}
;

module.exports = chainGenerator;
